gprofiler/cloud_metadata.py

```diff
@@ -0,0 +1,161 @@
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+
+import requests
+from requests import Response
+
+from gprofiler.exceptions import BadResponseCode
+from gprofiler.log import get_logger_adapter
+
+METADATA_REQUEST_TIMEOUT = 5
+
+logger = get_logger_adapter(__name__)
+
+
+@dataclass
+class InstanceMetadataBase:
+    provider: str
+
+
+@dataclass()
+class AwsInstanceMetadata(InstanceMetadataBase):
+    region: str
+    zone: str
+    instance_type: str
+    life_cycle: str
+    account_id: str
+    image_id: str
+    instance_id: str
+
+
+@dataclass
+class GcpInstanceMetadata(InstanceMetadataBase):
+    zone: str
+    instance_type: str
+    preempted: bool
+    preemptible: bool
+    instance_id: str
+    image_id: str
+    name: str
+
+
+@dataclass
+class AzureInstanceMetadata(InstanceMetadataBase):
+    instance_type: str
+    zone: str
+    region: str
+    subscription_id: str
+    resource_group_name: str
+    resource_id: str
+    instance_id: str
+    name: str
+    image_info: Optional[Dict[str, str]]
+
+
+def get_aws_metadata() -> Optional[AwsInstanceMetadata]:
+    metadata_response = send_request("http://169.254.169.254/latest/dynamic/instance-identity/document")
+    life_cycle_response = send_request("http://169.254.169.254/latest/meta-data/instance-life-cycle")
+    if life_cycle_response is None or metadata_response is None:
```

Comment:
Why either? If any succeeds, isn't it AWS?

Reply:
Or both must always succeed anyway

Reply:
I don't want to send a partial response. If one of them fails, it will also fail on the other cloud types and therefore it will log the exceptions

Reply:
k